azure-kusto-ingest/azure/kusto/ingest/ingest_client.py

```diff
@@ -108,3 +125,32 @@ def ingest_from_blob(self, blob_descriptor: BlobDescriptor, ingestion_properties
         content = ingestion_blob_info_json
         queue_client = queue_service.get_queue_client(queue=random_queue.object_name, message_encode_policy=TextBase64EncodePolicy())
         queue_client.send_message(content=content)
+
+    def validate_endpoint_service_type(self):
+        if not hasattr(self, "_endpoint_service_type") or len(self._endpoint_service_type) == 0 or self._endpoint_service_type.isspace():
+            self._endpoint_service_type = self._retrieve_service_type()
+
+        if not self._EXPECTED_SERVICE_TYPE == self._endpoint_service_type:
+            message = self._WRONG_ENDPOINT_MESSAGE.format(self._EXPECTED_SERVICE_TYPE, self._endpoint_service_type)
+            has_endpoint = True
+            if not hasattr(self, "_suggested_endpoint_uri") or len(self._suggested_endpoint_uri) == 0 or self._suggested_endpoint_uri.isspace():
+                self._suggested_endpoint_uri = self._generate_endpoint_suggestion(self._connection_datasource)
+                if not hasattr(self, "_suggested_endpoint_uri") or len(self._suggested_endpoint_uri) == 0 or self._suggested_endpoint_uri.isspace():
```

Comment:
why double check? remove has_endpoint and check:
if self._suggested_endpoint_uri

Reply:
Since this is a fault flow which needs fixing by the user, I see no point in caching the message creation.

Reply:
Because this flow is slightly more complex than it seems. The invisible case is if _suggested_endpoint_uri is defined in the first place.

Reply:
But _generate_endpoint_suggestion returns an empty string if failed so anyway in line 138 you have this variable set.

Reply:
Yes, line 137 can be shortened. I kept it there because I think the consistency among the 3 lines that do the same thing (133, 135, 137) is more maintainable (easier to read and understand).

But since you think it's better to remove the hasattr from the check in the final line I'll change it.